File: backend/app/schemas/workflow.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import UUID

class WorkflowCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: Optional[str] = None
    definition: Optional[Dict[str, Any]] = Field(default_factory=dict)

class WorkflowUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    definition: Optional[Dict[str, Any]] = None

class WorkflowResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    name: str
    description: Optional[str]
    status: str
    version: int
    definition: Dict[str, Any]
    created_by: Optional[UUID]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

class WorkflowExecutionResponse(BaseModel):
    id: UUID
    workflow_id: UUID
    tenant_id: UUID
    status: str
    variables: Dict[str, Any]
    started_at: datetime
    completed_at: Optional[datetime]
    
    class Config:
        from_attributes = True

File: frontend/src/types/workflow.ts
export interface Workflow {
  id: string;
  name: string;
  description: string;
  version: number;
  status: WorkflowStatus;
  definition: WorkflowDefinition;
  schedule?: WorkflowSchedule;
  tenantId: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export type WorkflowStatus = 'draft' | 'active' | 'paused' | 'error' | 'archived';

export interface WorkflowDefinition {
  nodes: ProcessGraphNode[];
  edges: ProcessGraphEdge[];
  variables: WorkflowVariable[];
  triggers: WorkflowTrigger[];
}

export interface ProcessGraphNode {
  id: string;
  type: NodeType;
  position: { x: number; y: number };
  data: NodeData;
}

export type NodeType = 'start' | 'end' | 'task' | 'decision' | 'human_task' | 'subprocess' | 'ai_skill';

export interface NodeData {
  label: string;
  description?: string;
  config: Record<string, any>;
  inputMapping?: Record<string, string>;
  outputMapping?: Record<string, string>;
  timeout?: number;
  retryConfig?: RetryConfig;
}

export interface ProcessGraphEdge {
  id: string;
  source: string;
  target: string;
  label?: string;
  condition?: string;
}

export interface WorkflowVariable {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object' | 'array';
  defaultValue?: any;
  required: boolean;
}

export interface WorkflowTrigger {
  type: 'schedule' | 'webhook' | 'event' | 'manual';
  config: Record<string, any>;
}

export interface WorkflowSchedule {
  cron: string;
  timezone: string;
  enabled: boolean;
}

export interface RetryConfig {
  maxAttempts: number;
  backoffMultiplier: number;
  initialDelaySeconds: number;
}

export interface WorkflowExecution {
  id: string;
  workflowId: string;
  status: 'running' | 'completed' | 'failed' | 'cancelled';
  startedAt: string;
  completedAt?: string;
  currentNode?: string;
  variables: Record<string, any>;
  events: ExecutionEvent[];
}

export interface ExecutionEvent {
  id: string;
  type: string;
  nodeId?: string;
  data: Record<string, any>;
  timestamp: string;
}